Added selected and deselected text properties to OCKCareCardButton

moppymopperson opened the pull request:
This pull request probably needs some extra input before it's ready to merge. It's purpose is to add an optional label below `OCKCareCardButton`, but there is room for improvement in the way that data is passed to the cells. 

<img src="https://user-images.githubusercontent.com/22581112/35184628-a4e45666-fe3b-11e7-9676-ace5334b5b2d.png" width="450">

Currently I am passing the text in through `userInfo`, but this doesn't seem like a good way to make it a standard feature. If you have a preference for how to get the data to the button objects in a less hacky way, I'd be happy to update this PR.

ninoguba commented:
This looks like a useful addition.  I suggest you pass the "info" thru the OCKCareCardViewControllerDelegate protocol.  Either as a new parameter to this existing function or add a new function.

http://carekit.org/docs/Protocols/OCKCareCardViewControllerDelegate.html#//api/name/careCardViewController:didSelectButtonWithInterventionEvent:

moppymopperson commented:
@ninoguba Thanks for the idea. I'm much more happy with this setup.

akshay-yadav-apple commented on CareKit/CareCard/OCKCareCardButton.h:
```diff
@@ -1,5 +1,6 @@
 /*
  Copyright (c) 2016, Apple Inc. All rights reserved.
+ Copyright (c) 2018, Erik Hornberger. All rights reserved.s
```
Typo.

akshay-yadav-apple commented on CareKit/CareCard/OCKCareCardButton.m:
```diff
@@ -53,6 +55,16 @@ - (void)drawRect:(CGRect)rect {
         UIRectFill(_circleLayer.frame);
     }
+    if (!_label) {
+        CGRect labelRect = CGRectMake(0, ButtonSize/2 + 8, ButtonSize, ButtonSize);
```
I think creating a static const of value 8, with an explanatory variable name will be better.

akshay-yadav-apple commented on CareKit/CareCard/OCKCareCardButton.m:
```diff
@@ -53,6 +55,16 @@ - (void)drawRect:(CGRect)rect {
+    if (!_label) {
+        CGRect labelRect = CGRectMake(0, ButtonSize/2 + 8, ButtonSize, ButtonSize);
+        _label = [[UILabel alloc] initWithFrame:labelRect];
+        _label.font = [UIFont systemFontOfSize:9 weight:300];
```
Can we use any UIKIT_EXTERN const for a UIFontWeight instead of 300. I think 'heavy' looks good.

akshay-yadav-apple commented on CareKit/CareCard/OCKCareCardButton.m:
```diff
@@ -68,6 +80,10 @@ - (void)updateFillColorForSelection:(BOOL)selection {
     }
 }
 
+- (void)updateTextForSelection:(BOOL)selection {
```
extraneous parameter 'selection' if checking against self.isSelected.

akshay-yadav-apple commented on CareKit/CareCard/OCKCareCardTableViewCell.h:
```diff
@@ -47,6 +48,10 @@ NS_ASSUME_NONNULL_BEGIN
 
 - (void)careCardTableViewCell:(OCKCareCardTableViewCell *)cell didSelectInterventionActivity:(OCKCarePlanActivity *)activity;
 
+- (nullable NSString *)careCardTableViewCell:(OCKCareCardTableViewCell *)cell selectedStateTextForInterventionActivity: (OCKCarePlanEvent *)interventionActivity atIndex:(NSInteger) index;
```
parameter name 'interventionActivity' of type OCKCarePlanEvent might be misleading.

akshay-yadav-apple commented on CareKit/CareCard/OCKCareCardTableViewCell.h:
```diff
@@ -47,6 +48,10 @@ NS_ASSUME_NONNULL_BEGIN
 
+- (nullable NSString *)careCardTableViewCell:(OCKCareCardTableViewCell *)cell selectedStateTextForInterventionActivity: (OCKCarePlanEvent *)interventionActivity atIndex:(NSInteger) index;
+
+- (nullable NSString *)careCardTableViewCell:(OCKCareCardTableViewCell *)cell deselectedStateTextForInterventionActivity: (OCKCarePlanEvent *)interventionActivity atIndex:(NSInteger) index;
```
Review same as selectedStateTextForInterventionActivity

akshay-yadav-apple commented on CareKit/CareCard/OCKCareCardViewController.h:
```diff
@@ -117,6 +117,24 @@ NS_ASSUME_NONNULL_BEGIN
+ @param interventionActivity        The `OCKCarePlanEvent` that the button represents
+ @param index                       The index of the `OCKCareCardButton` within the cell
+ */
+- (nullable NSString *)careCardViewController:(OCKCareCardViewController *)viewController selectedStateTextForInterventionActivity: (OCKCarePlanEvent *)interventionActivity atButtonIndex: (NSInteger)index;
```
parameter name 'interventionActivity' of type OCKCarePlanEvent might be misleading.

akshay-yadav-apple commented on CareKit/CareCard/OCKCareCardViewController.m:
```diff
@@ -627,6 +627,19 @@ - (void)careCardTableViewCell:(OCKCareCardTableViewCell *)cell didSelectInterven
 }
 
+- (nullable NSString*) careCardTableViewCell:(OCKCareCardTableViewCell *)cell selectedStateTextForInterventionActivity:(OCKCarePlanEvent *)interventionActivity atIndex:(NSInteger)index {
+    if ([self.delegate respondsToSelector:@selector(careCardTableViewCell:selectedStateTextForInterventionActivity:atIndex:)]){
```
checking against different selector

akshay-yadav-apple commented on CareKit/CareCard/OCKCareCardViewController.m:
```diff
@@ -627,6 +627,19 @@ - (void)careCardTableViewCell:(OCKCareCardTableViewCell *)cell didSelectInterven
+}
+
+- (nullable NSString*) careCardTableViewCell:(OCKCareCardTableViewCell *)cell deselectedStateTextForInterventionActivity:(OCKCarePlanEvent *)interventionActivity atIndex:(NSInteger)index {
+    if ([self.delegate respondsToSelector:@selector(careCardTableViewCell:deselectedStateTextForInterventionActivity:atIndex:)]){
```
checking against different selector

akshay-yadav-apple commented on CareKit/CareContents/OCKCareContentsViewController.h:
```diff
@@ -137,6 +140,24 @@ NS_ASSUME_NONNULL_BEGIN
+ @param interventionActivity        The `OCKCarePlanEvent` that the button represents
+ @param index                       The index of the `OCKCareCardButton` within the cell
+ */
+- (nullable NSString *)careContentsViewController:(OCKCareContentsViewController *)viewController selectedStateTextForInterventionActivity: (OCKCarePlanEvent *)interventionActivity atButtonIndex: (NSInteger)index;
```
parameter name 'interventionActivity' of type OCKCarePlanEvent might be misleading.

akshay-yadav-apple commented on CareKit/CareContents/OCKCareContentsViewController.m:
```diff
@@ -908,6 +909,19 @@ - (void)careCardTableViewCell:(OCKCareCardTableViewCell *)cell didSelectInterven
 }
 
+- (nullable NSString*)careCardTableViewCell:(OCKCareCardTableViewCell *)cell selectedStateTextForInterventionActivity:(OCKCarePlanEvent *)interventionActivity atIndex:(NSInteger)index {
+    if ([self.delegate respondsToSelector:@selector(careCardTableViewCell:selectedStateTextForInterventionActivity:atIndex:)]){
```
checking against different selector

akshay-yadav-apple commented on CareKit/CareContents/OCKCareContentsViewController.m:
```diff
@@ -908,6 +909,19 @@ - (void)careCardTableViewCell:(OCKCareCardTableViewCell *)cell didSelectInterven
+}
+
+- (nullable NSString*)careCardTableViewCell:(OCKCareCardTableViewCell *)cell deselectedStateTextForInterventionActivity:(OCKCarePlanEvent *)interventionActivity atIndex:(NSInteger)index {
+    if ([self.delegate respondsToSelector:@selector(careCardTableViewCell:deselectedStateTextForInterventionActivity:atIndex:)]){
```
checking against different selector

moppymopperson commented:
@akshay-yadav-apple Thanks for the feedback! I’ll make all those changes soon.

moppymopperson commented on CareKit/CareCard/OCKCareCardTableViewCell.m:
```diff
@@ -109,13 +109,13 @@ - (void)prepareView {
                   forControlEvents:UIControlEventTouchUpInside];
         
         if (self.delegate &&
```
Here we're checking if the delegate is set and if it responds to a selector, but to the best of my knowledge there are presently no cases where the delegate is not set or the method in question is not implemented.

The two places this cell is used are `OCKCareCardViewController` and `OCKCareContentsViewController`, and in both cases the cell delegate is set in the cellForIndexPath table view data source methods. Moreover, as of this PR both have a default implementation of the selectors in question that just returns `NULL`. 

Do you think it's better to leave these two checks in place incase `OCKCareCardButton` ever becomes a public class, or should we take out the checks since they are superfluous?

akshay-yadav-apple replied:
@moppymopperson, I agree, it is always a good idea to check if delegate exists, and if delegate responds to the selector, before calling it.

moppymopperson replied:
@akshay-yadav-apple Glad to hear that, thanks for the feedback.

moppymopperson commented:
All changes have been made. Additionally, I added some spacing between the rows of buttons so that the labels will appear spaced nicely between rows when there are more than 8 buttons. I also added a some default text to the sample app for your convenience when reviewing.

<img src="https://user-images.githubusercontent.com/22581112/35468582-78be5a90-0364-11e8-89fa-000812e69b46.png" width="450">

akshay-yadav-apple commented:
Hey @moppymopperson,
Before we go ahead and merge the PR, can you please take a look at, and resolve the merge conflicts.

moppymopperson commented:
Sure, I rebased onto to master instead of merging to prevent unrelated commits from getting mixed into this PR.

moppymopperson commented:
@akshay-yadav-apple Any updates on this? Once this is merged I think I can finally go back to using the official version of CareKit instead of my own fork.

akshay-yadav-apple commented:
Hey @moppymopperson,
We are still running some tests, and still reviewing the PR. 
We will get back to you very soon with the next step.

moppymopperson commented:
@akshay-yadav-apple Anything I can help with to move this forward?

akshay-yadav-apple commented:
Hey @moppymopperson,

The PR looks good but we have a few questions. 
Can you please confirm what the contributions are from the following parties:
+ Copyright (c) 2017, EExT LLC. All rights reserved.
+ Copyright (c) 2018, Swift愛好会. All rights reserved.

moppymopperson commented:
@akshay-yadav-apple EExT is a small company that a friend and I created to develop apps under. Most of the changes in this PR were made while I was on the clock for EExT, so I credited the company. All the work related to the UI was done under this scope. At this point I was just grabbing data out of the `userInfo` dictionary instead of using a proper delegate setup.

Swift愛好会 is an informal dev meet-up here in Japan for people interested in Swift. They hosted a two day retreat for working on hacks and personal projects, and it was there that I implemented all the delegate logic that allows the user to set the text without having to use some kind of add hoc `userInfo` hack.

If either of those are problematic, you're welcome to just replace the two of them with my name as in past PR's. I'm really not picky either way - whatever is easiest for you all.

moppymopperson commented:
@akshay-yadav-apple Any updates on this? I'm happy to modify the copyright if it gets this across the finish line.

moppymopperson commented:
@akshay-yadav-apple This has been sitting in the queue for awhile. Have you found a problem with it or something?

akshay-yadav-apple commented:
Hey @moppymopperson,
I am very sorry that this PR had to stay idle for so long. This PR looks like a very useful addition, and I am trying to think of as many edge cases we can cover before merging this in. 
Since the label can only be as long as the width of the button, I am concerned there isn't much text that can be used with the current implementation. Longest string that can be added could be around 5-7 characters long. Do you think there should be an assertion for the text size, or should the next button be moved to the next line when the text grows bigger? I also feel like we should not have a strict 9pt. font size. Let me know what are your thoughts around the size of the label, and we can surely take the PR forward.

akshay-yadav-apple commented:
@moppymopperson
Good idea!
I also think if you're considering changing the constraints, the label could be muti-line, and the width of the label should be around one-third of the cell, so that at least 3 buttons (with labels) can fit in one row, and if the text is too long, it'll keep moving to next lines and keep pushing next set of buttons further down. what do you think about that?

moppymopperson commented:
@akshay-yadav-apple Before I begin trying to implement that, let me ask a few questions to ensure I understand your intention properly. Are you imagining a fixed width label, such that there will always be exactly 3 buttons per row? Or would the maximum width of the label be 1/3 the cell width, such that there may be more than 3 buttons per row if the label text is short?

srinathtm-apple commented:
Closing out PRs that are no longer relevant due to the CareKit 2.0 beta update.
